man/systemd-resolve.xml

```diff
@@ -369,6 +412,82 @@
+    compatibility mode. It accepts mostly the same arguments and pushes all data into
+    <citerefentry><refentrytitle>systemd-resolved.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>,
+    similar to how <option>--set-dns=</option> and <option>--set-domain=</option> operate. Note that
+    <command>systemd-resolved.service</command> is the only supported backend, which is different from other
+    implementations of this command. Note that not all operations supported by other implementations are supported
+    natively. Specifically:</para>
+
+    <variablelist>
+      <varlistentry>
+        <term><option>-a</option></term>
+        <listitem><para>Registers per-interface DNS configuration data with
+        <command>systemd-resolved</command>. Expects a network interface name as only command line argument. Reads
+        <citerefentry><refentrytitle>resolv.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry> compatible DNS
+        configuration data from its standard input. Relevant fields are <literal>nameserver</literal> and
+        <literal>domain</literal>/<literal>search</literal>. This command is mostly identical to invoking
+        <command>systemd-resolve</command> with a combination of <option>--set-dns=</option> and
+        <option>--set-domain=</option>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>-d</option></term>
+        <listitem><para>Unregisters per-interface DNS configuration data with <command>systemd-resolved</command>. This
+        command is mostly identical to invoking <command>systemd-resolve</command> with
+        <option>--revert</option>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>-f</option></term>
+
+        <listitem><para>When specified <option>-a</option> and <option>-d</option> will not complain about missing
+        network interfaces and will silently execute no operation in that case.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>-x</option></term>
+
+        <listitem><para>This switch for "exclusive" operation is supported only partially. It is mapped to an
+        additional configured search domain of <literal>~.</literal> — i.e. ensures that DNS traffic is preferably
+        routed to the DNS servers on this interface, unless there are other, more specific domains configured on other
+        interfaces.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>-m</option></term>
+        <term><option>-p</option></term>
+
+        <listitem><para>These switches are not supported and are silently ignored.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>-u</option></term>
+        <term><option>-I</option></term>
+        <term><option>-i</option></term>
+        <term><option>-l</option></term>
```

Comment:
Would you consider supporting `-l [interface]`? The information and logic to support it _is_ easily available, and so it wouldn't be cumbersome to support. This is used by `wg-quick(8)` to (optionally) save the active runtime configuration when deleting the interface.

Reply:
I think providing the API for 3rd party tools to hook into is priority, and I figure that's mostly -a and -d. OTOH -l is mostly user-facing, and we do have "systemd-resolve --status" covering that already (and without the skewedness because resolved's and resolvconf's concept don't match 100%). 

That said, I think we can totally add support for that later on, if people want that (and apparently they do, as you have just shown). But I think it's just the cherry on top, and not not necessary to make the basic functionality work.

Or to say this differently: a PR adding support for that on top of this once it's merged would be very welcome!

Reply:
Indeed the other program supports many things too, but the point of having resolvconf in the first place is to maintain compatibility with scripts, just like `shutdown` and the rest. `wg-quick(8)` won't be growing support for a cornucopia of different DNS management invocations, preferring instead to use a single API (resolvconf). It's just a shell script after all.

You can see its usage of `-l` here: https://git.zx2c4.com/WireGuard/tree/src/tools/wg-quick.bash#n194

> But I think it's just the cherry on top, and not necessary to make the basic functionality work.

It's certainly required to make `wg-quick(8)` work, rather than being a mere cherry.

Reply:
note that "-l" is only supported by openresolv, btw, it's not available in the original Debian implementation.

Reply:
Hmm, that script uses `-m 0 -x` btw, which isn't in the debian implementation. It appears the script only works with openresolv.

What confuses me a bit is this line in that script:

```resolvconf -a "tun.$INTERFACE" -m 0 -x```

The interface argument is reversed: According to the manpages you are supposed to specified the interface first, and the protocol second. i.e. it should be `$INTERFACE.tun` and not `tun.$INTERFACE`. (Oh, and I think it would be more expressive if you'd use "wguard or so instead of "tun" as protocol ID).

Reply:
See these two links regarding Debian's own implementation of the script, and the one from "openresolv". Debian packages both interestingly:

https://manpages.debian.org/stretch/resolvconf/resolvconf.8.en.html
https://manpages.debian.org/stretch/openresolv/resolvconf.8.en.html

The FreeBSD version is even different:

https://www.freebsd.org/cgi/man.cgi?query=resolvconf&sektion=8

Reply:
I figure if you want to be compatible with Debian's, openresolv's and FreeBSD's implementation you'd have to rewrite your command line to use this syntax:

```sh
IF_METRIC=0 IF_EXCLUSIVE=1 resolvconf -a $INTERFACE.wireguard < …
```
That line also does the right thing with the implementation provided by this PR, as it also honours the IF_EXCLUSIVE command line arguments.

And I fear that the "-l" thing you can't use if you care about general Debian compatibility.

Reply:
Actually those arguments are silently ignored, which is why, I assume, they're also ignored on systemd. This allows us to keep compat on both.

We're then doing the tun. prefix hack do that the interface is ordered on top, as Debian's uses a fixed order based on interface prefix. This simulates the -m 0 behavior, though we're out on -x. I hadn't seen those environment variables before and I wonder if they're new. I'll take a look and add those in, since they importantly won't do harm on other implementations.

The take away is that since Debian turns unknown instructions into no-ops, we can support multiple ones at the same time.

Reply:
FreeBSD uses openresolv, by the way.

Reply:
> Actually those arguments are silently ignored, which is why, I assume, they're also ignored on systemd.

are they? i just checked out the resolvconf upstream git repo, and if i run the command there with -x it will say "resolvconf: Error: Command not recognized" and refuse operation. I don't think -x and -m works on the Debian resolvconv. Hence if you want to communicate the metric or exclusive mode to resolvconf where it's supported then you have to go the env var way

Reply:
> We're then doing the tun. prefix hack do that the interface is ordered on top, as Debian's uses a fixed order based on interface prefix. 

Uh, but that's against the documentation, and I am pretty sure will break various backend implementations that actually care that the interface name you specify is valid. I mean, the original resolvconf exists precisely to allow multiple backends, no? If you use it on top of classic bind, then the interface name needs to be parsable properly, as one exaple.

The compat implementation of this PR also expects that the interface name is valid. It will generate an error if it isn't.

Reply:
> FreeBSD uses openresolv, by the way.

I think it's based on it, yes, but it deviates, has new options and the file locations are all different.

Reply:
> I think it's based on it, yes,

That man page you linked explicitly mentions openresolv, anyhow.

> are they? i just checked out the resolvconf upstream git repo, and if i run the command there 

Wow, so maybe Debian changed in their latest git repo? In all currently released versions it's silently ignored. I had opened this -- https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=860564 -- a while ago, hoping they'd switch to using openresolv. Maybe instead they just decided to extend their own homegrown monstrosity with an incompatible interface (those environment variables you mentioned). If so, that's quite disturbing for the current state of affairs. I'll check into it. Maybe @dkg can shed some light too.

Reply:
> are they? i just checked out the resolvconf upstream git repo, and if i run the command there

I'm reading this source: https://anonscm.debian.org/cgit/resolvconf/resolvconf.git/tree/bin/resolvconf

The key is to make -x the second argument, not the first. Then you avoid the error message and have future compatibility with openresolv.

Reply:
> The compat implementation of this PR also expects that the interface name is valid. It will generate an error if it isn't.

Ahh, okay, that could be problematic, then, unless what you're saying about `IF_EXCLUSIVE` is actually true in current versions, in which case, I wouldn't need the `tun.` prefix hack.

Reply:
> note that "-l" is only supported by openresolv, btw, it's not available in the original Debian implementation.

https://git.zx2c4.com/WireGuard/commit/?id=6ed55dec679b2aaa36b4170632cdc510475fc310

So, it might be a good idea for systemd to support the modern interface for this instead of the old Debian cruft.